tests/integration/test_alert_rules.py

```diff
@@ -0,0 +1,117 @@
+
+
+@pytest.mark.abort_on_fail
+async def test_no_rules_files(ops_test):
+    """Check that there are no alert rules yet."""
+    logger.debug("deploy charms from charmhub")
+
+    rules_before_relation = await loki_rules(ops_test, app_name)
+    assert len(rules_before_relation) == 0
+
+
+@pytest.mark.abort_on_fail
+async def test_first_relation_one_alert_rule(ops_test):
+    await asyncio.gather(
+        ops_test.model.deploy(
+            "ch:cos-configuration-k8s",
+            application_name=rules_app,
+            channel="edge",
+            config={
+                "git_repo": "https://github.com/canonical/loki-k8s-operator",
+                "git_branch": "main",
+                "loki_alert_rules_path": "tests/sample_rule_files/free-standing",
+            },
+        ),
+    )
+
+    # force an update, just in case the files showed up on disk after the last hook fired
+    action = await ops_test.model.applications[rules_app].units[0].run_action("sync-now")
+    await action.wait()
+
+    # form a relation between loki and the app that provides rules
+    await ops_test.model.add_relation(app_name, rules_app)
+
+    async with IPAddressWorkaround(ops_test):
+        await ops_test.model.wait_for_idle(
+            apps=[app_name, rules_app], status="active", timeout=1000
+        )
+
+    rules_after_relation = await loki_rules(ops_test, app_name)
+    assert len(rules_after_relation) == 1
+
+
+@pytest.mark.abort_on_fail
+async def test_second_relation_second_alert_rule(ops_test):
+    await asyncio.gather(
+        ops_test.model.deploy(
+            "ch:cos-configuration-k8s",
+            application_name=rules_app2,
+            channel="edge",
+            config={
+                "git_repo": "https://github.com/canonical/loki-k8s-operator",
+                "git_branch": "main",
+                "loki_alert_rules_path": "tests/sample_rule_files/free-standing",
+            },
+        ),
+    )
+
+    # force an update, just in case the files showed up on disk after the last hook fired
+    action = await ops_test.model.applications[rules_app2].units[0].run_action("sync-now")
+    await action.wait()
+
+    # form a relation between loki and the app that provides rules
+    await ops_test.model.add_relation(app_name, rules_app2)
+
+    async with IPAddressWorkaround(ops_test):
+        await ops_test.model.wait_for_idle(
+            apps=[app_name, rules_app2], status="active", timeout=1000
+        )
+
+    rules_after_relation2 = await loki_rules(ops_test, app_name)
+    assert len(rules_after_relation2) == 2
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_app_one_alert_rules_is_reteined(ops_test):
+    await ops_test.model.applications[rules_app].remove()
+    await ops_test.model.block_until(lambda: rules_app not in ops_test.model.applications)
+
+    rules_after_delete_relation2 = await loki_rules(ops_test, app_name)
+    assert len(rules_after_delete_relation2) == 1
```

Comment:
If instead you remove `rules_app2` then you could easily have a stronger assertion:
```suggestion
    assert rules_after_delete_relation2 == rules_after_relation
```

Reply:
Will do it in: https://github.com/canonical/loki-k8s-operator/issues/103